docs/rfcs/rfc03.md

```diff
@@ -0,0 +1,127 @@
+# RFC02: Caching and vendoring
+
+## Background
+
+Jsonnet goes way out of its way to ensure that all the imports are well known statically, and that the evaluation is fully reproducible
+and cacheable by providing side-effect free evaluation and avoiding computed imports.
+
+Kubecfg adds the ability to directly import external dependencies. This is meant to be used for importing immutable external files,
+usually links to some resource at some specific git sha. It also [supports](https://github.com/kubecfg/kubecfg/pull/218) explicitly enforcing a fingerprint check.
+
+However, fetching external dependencies comes with problems:
+
+* downloding external dependencies may be slow and annoying
+* external files may disappear permanently or temporarily
+
+## Proposal
+
+We propose adding a download cache that will store a local copy of files downloaded from remote sources.
+
+The cache directory structure, is suitable for checking-in the cache into the current VCS repo (if the user so chooses).
+
+### Cache structure
+
+Let's consider this jsonnet source file:
+
+```jsonnet
+import 'https://github.com/kubecfg/k8s-libsonnet/raw/144e62ccbdc9f7d4f8bccc14f33224cbf63e5185/k8s.libsonnet'
+```
+
+The cache directory can be set via the `--cache-dir` flag. By default its `<cachebase>/kubecfg`, where `<cachebase>` is determined by the [os.UserCacheDir](https://pkg.go.dev/os@go1.20.2#UserCacheDir).
+
+
+After the first execution, kubecfg will create this directory structure
+
+```console
+$ tree ~/Library/Caches/kubecfg
+/Users/mkm/Library/Caches/kubecfg
+└── https
+    └── github.com
+        └── kubecfg
+            └── k8s-libsonnet
+                └── raw
+                    └── 144e62ccbdc9f7d4f8bccc14f33224cbf63e5185
+                        └── k8s.libsonnet
+
+
+### Multiple cache sources
+
+(optional, advanced)
+
+The user can define multiple cache directories.
+They will act as a "search path".
+If a file is not found in the first directory, the next one will be tried until the cache directories are exhausted.
+Only the first cache directory is used to write new files to.
+
+Example:
+
+```bash
+kubecfg --cache-dir ./vendor --cache-dir ~/.cache/kubecfg show ....
+```
+
+Open questions:
+
+* should we always put the default cache at the end of the cache search path?
```

Comment:
I think we need a way to say "no cache".  I think we should have a `--no-user-cache` flag (defaults to _using_ a homedir cache), which internally just adjusts whether we put the default homedir cache at the end of the search path.

In particular this means we can do weird things like `--cache-dir ./vendor --no-user-cache`, and it still sort of makes sense.